File: AlgorithmReview/Programmers/Lv2/뒤에 있는 큰 수 찾기.swift
//
//  뒤에 있는 큰 수 찾기.swift
//  AlgorithmReview
//
//  Created by 박소윤 on 2023/05/23.
//

import Foundation

func solution(_ numbers:[Int]) -> [Int] {
    
    var ans = [-1]
    var stack = [numbers.last!]
    for i in stride(from: numbers.count-2, to: -1, by: -1) {
        
        while !stack.isEmpty && stack.last! <= numbers[i] {
            stack.removeLast()
        }
        
        if stack.isEmpty {
            ans.append(-1)
        }
        else {
            ans.append(stack.last!)
        }
        
        stack.append(numbers[i])
    }
    
    return ans.reversed()
}

File: Python/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    
    answer = [-1] * len(numbers)
    stack = [numbers[-1]]
    
    for i in range(len(numbers)-2, -1, -1):
        
        while len(stack) > 0 and stack[-1] <= numbers[i]:
            del stack[-1]
        
        if len(stack) != 0:
            answer[i] = stack[-1]
        
        stack.append(numbers[i])
    
    return answer
